apps/E2E/src/Checkbox/pages/CheckboxPageObject.ts

```diff
@@ -21,45 +15,34 @@ class CheckboxPageObject extends BasePage {
     return await (await this._primaryComponent).isSelected();
   }
 
-  async waitForCheckboxChecked(timeout?: number): Promise<void> {
-    await browser.waitUntil(async () => await this.isCheckboxChecked(), {
-      timeout: timeout ?? this.waitForUiEvent,
-      timeoutMsg: 'The Checkbox was not toggled correctly',
-      interval: 1000,
-    });
+  /* Waits for the checkbox to be checked or unchecked if the new state is true or false. Returns true if the
+   * checkbox toggled to the new state. */
+  async waitForCheckboxToggle(newState: boolean, errorMessage?: string): Promise<boolean> {
+    if ((await this.isCheckboxChecked()) !== newState) {
+      await this.waitForCondition(async () => (await this.isCheckboxChecked()) === newState, errorMessage);
+    }
+    return (await this.isCheckboxChecked()) === newState;
   }
 
-  /* Useful in beforeEach() hook to reset the checkbox before every test */
-  async toggleCheckboxToUnchecked(): Promise<void> {
-    if (await this.isCheckboxChecked()) {
-      await (await this._primaryComponent).click();
+  /* Toggles the checkbox to the checked if newState == true or unchecked if newState == false. */
+  async toggleCheckbox(newState: boolean) {
+    if ((await this.isCheckboxChecked()) !== newState) {
+      await this.click(this._primaryComponent);
+      await this.waitForCheckboxToggle(
+        newState,
+        `Clicked the primary checkbox to turn it ${newState ? 'on' : 'off'}, but it failed to change states.`,
+      );
     }
   }
 
+  /* The onChange callback for the checkbox shows a hidden Text component near the checkmark. This checks if the text
```

Comment:
I'd recommend providing more info on why we need the text in the first place (that's the confusing part about this function).

"Unlike snapshot testing, we can't spy on functions to determine if they get called or not. So, in order to determine if the onChange() callback gets fired, we show/hide a text component as the onChange() callback is fired. Then, we can check to see if this text exists to determine if the onChange() fired."